### ISSUE TITLE
ThreadLocal causing errors in StdSet

### ISSUE THREAD

**Issue by jgFages** [chocoteam/choco-solver#519]
ThreadLocal leads to errors : See SetVarImplTest.testStructures (in next commit)
=> I comment it until a better solution is found

Also minor bug detected in Set_Std_Swap2.remove() (wrong return, see BacktrackableSetTest)
=> I fix it

**Comment by cprudhom** [comment]
I let it open otherwise we will never fix it

**Comment by cprudhom** [comment]
I found how to fix the issue, but I'm wondering if we can simply remove the `ListOp` object and pool manager with lambda:
Instead of: 
```java
ListOP op = operationPoolGC.getE();
if (op == null) {
     op = new ListOP();
}
op.set(set, iter.nextInt(), ADD);
```
do this:
```java
environment.save(() -> set.add(iter.next()));
```

The following micro-bench:
```java
SetFactory.HARD_CODED = false;                                       
Model m = new Model();                                               
IEnvironment environment = m.getEnvironment();                       
StdSet set = new StdSet(environment, SetFactory.makeBipartiteSet(0));
boolean add = true;                                                  
for(int i = 0 ; i< 99999999; i++){                                   
	environment.worldPush();                                         
	add &=set.add(0);                                                
	environment.worldPop();                                          
}                                                                    
System.out.printf("%s", add);                                        
```
goes from 8.3 seconds to 7.6.
I didn't evaluate the memory footprint, but at least it is faster.

**Comment by jgFages** [comment]
The pool is mainly used to reduce memory leaks, as for decisions objects. 

````
IntDecision d = intDecisionPool.getE();
if (d == null) {
     d = new IntDecision(intDecisionPool);
}
d.set(var, value, dop);
````

Would you suggest to remove the pool for them too?

**Comment by cprudhom** [comment]
In the case of `StdSet`, `ListOp` is present to encapsulate a reverse operation, its existence (and the usage of a pool manager) is only justified to limit object creation and to favor object recycling (which I strongly encourage :) .
But, `ListOp` can be converted into lambdas, since `IOperation` is an interface with only one method, this provides better readability and avoid using a shared pool manager.
The operation itself is very simple, and can certainly be well managed by the JVM as a lambda (well I suppose).
So, to me it is an option to consider replacing `LisOp` by lambdas, and consequently, remove the pool manager, if the memory leaks is limited and the performance is better.
We can also consider `Propagator.state` operations as lambdas too (their evaluation could also be easier, since they are widely used).

Decision object is more complex and cannot be replaced by lambda, they certainly consume more GC time if not recycled, and it is easy to see that each backtrack release one decision which can be automatically reuse. So I think we should keep them as is.

**Comment by jgFages** [comment]
Thank you for the explanation!
I will try to do a crash test on memory and I let you know!

**Comment by jgFages** [comment]
When doing this in BacktrackableSetTest:
````
    @Test(groups = "10s", timeOut=60000)
    public void memoryCrashTest() {
    	SetFactory.HARD_CODED = false;
        ISet set = create();
        assertTrue(set.add(2));
        IEnvironment environment = model.getEnvironment();
		for(int k=0;k<300;k++) {
			environment.worldPush();
			for (int i = 0; i < 1000; i++) set.add(i);
			environment.worldPop();
			assertTrue(set.size() == 1);
		}
    }
````

I obtain
`````
				|current	|lambda	|
______________________________________
BipartiteSet		| 0.1		| 0.1		|
BitSet			| 0.1		| 0.1		|
LinkedList		| 0.4		| 0.5		|
SmallBipartiteSet	| 10.6	| 10.7	|
``````
So I do not believe there is any significant difference in runtime, but it is definitely less code, which I like :-)

**Comment by cprudhom** [comment]
Ok, so let's go to that since having a PM attached to the thread leads to error in Travis  test suite, that I  was not able to reproduce locally.
I handle it.

**Comment by jgFages** [comment]
SetVarImplTest.testStructures passed on your machine? 
(I let you do it then)

**Comment by cprudhom** [comment]
The lambda version garbage collects a lot whereas the current version does not. 
But I can see some performance improvement, so I still believe this is a good idea to remove `ListOp` from `StdSet`.
I'm not sure concerning `Propagator.status` but I will evaluate it too (certainly, in another issue)

**Comment by cprudhom** [comment]
There is something crazy with `BipartiteSetTest.testClear`:
- when I run it alone, no problem,
- when it is run in the middle of the test suite, it creates 2Go of lambdas !

**Comment by cprudhom** [comment]
Ok, I think I found the issue, this is related to lazy execution of lambdas:
```java
while (iter.hasNext()) {
    environment.save(()->set.add(iter.nextInt()));
}
```java
never ends, but this ends:
while (iter.hasNext()) {
    int val = iter.nextInt();
    environment.save(()->set.add(val));
}
```

Indeed, since lambda is evaluated lazily, the `nextInt()` call will also be called later, and thus, the counter of the iterator is never incremented.

I'll fix this in my next commit, but may be this reveals a wrong usage of lambda...

**Comment by jgFages** [comment]
Thanks! We should keep this in mind!

**Comment by cprudhom** [comment]
I reduced  `memoryCrashTest` to smaller value because it seems that the way this test suite is declared does not handle time-out correctly and it has side effects on the other test.
Indeed, the `model` is shared between all tests and one goes wrong, some other may fail too, even if they do not test something similar.
In addition, they cannot be relaunched individually.
=> I revamped the test suite and pushed it in my next commit.
